File: node-src/ARControllerNFT.js
const ARToolkitNFT = require('./ARToolkitNFT.js')

class ARControllerNFT {
    constructor(width, height, cameraParam) {
        this.id = -1;

        this.width = width;
        this.height = height;

        // this is a replacement for ARCameraParam
        this.cameraParam = cameraParam;
        this.cameraId = -1;
        this.cameraLoaded = false;
        this.artoolkitNFT = global.artoolkitNFT;

        // to register observers as event listeners
        this.listeners = {};

        this.nftMarkerCount = 0;

        this.nftMarkers = {};

        this.markerFound = false;
        this.markerFoundTime = 0;

        this.transform_mat = new Float64Array(16);
        this.transformGL_RH = new Float64Array(16);
        this.marker_transform_mat = null;

        this.videoWidth = width;
        this.videoHeight = height;
        this.videoSize = this.videoWidth * this.videoHeight;

        this.framepointer = null;
        this.framesize = null;
        this.dataHeap = null;
        this.videoLuma = null;
        this.camera_mat = null;
        this.videoLumaPointer = null;
    }

    process(image) {
        var result = this.detectMarker(image);
        if (result != 0) {
            console.error("detectMarker error: " + result);
        }

        // get NFT markers
        var k, o;
        for (k in this.nftMarkers) {
            o = this.nftMarkers[k];
            o.inPrevious = o.inCurrent;
            o.inCurrent = false;
        }

        // detect NFT markers
        var nftMarkerCount = this.nftMarkerCount;
        this.detectNFTMarker();

        // in ms
        var MARKER_LOST_TIME = 200;

        for (var i = 0; i < nftMarkerCount; i++) {
            var nftMarkerInfo = this.getNFTMarker(i);
            var markerType = this.artoolkitNFT.NFT_MARKER;

            if (nftMarkerInfo.found) {
                this.markerFound = i;
                this.markerFoundTime = Date.now();

                var visible = this.trackNFTMarkerId(i);
                visible.matrix.set(nftMarkerInfo.pose);
                visible.inCurrent = true;
                this.transMatToGLMat(visible.matrix, this.transform_mat);
                this.transformGL_RH = this.arglCameraViewRHf(this.transform_mat);
                this.dispatchEvent({
                    name: 'getNFTMarker',
                    target: this,
                    data: {
                        index: i,
                        type: markerType,
                        marker: nftMarkerInfo,
                        matrix: this.transform_mat,
                        matrixGL_RH: this.transformGL_RH
                    }
                });
            } else if (this.markerFound === i) {
                // for now this marker found/lost events handling is for one marker at a time
                if ((Date.now() - this.markerFoundTime) <= MARKER_LOST_TIME) {
                    // not handling marker lost for less than specified time
                    return;
                }

                delete this.markerFound;

                this.dispatchEvent({
                    name: 'lostNFTMarker',
                    target: this,
                    data: {
                        index: i,
                        type: markerType,
                        marker: nftMarkerInfo,
                        matrix: this.transform_mat,
                        matrixGL_RH: this.transformGL_RH
                    }
                });
            }
        }
    };

    trackNFTMarkerId(id, markerWidth) {
        var obj = this.nftMarkers[id];
        if (!obj) {
            this.nftMarkers[id] = obj = {
                inPrevious: false,
                inCurrent: false,
                matrix: new Float64Array(12),
                matrixGL_RH: new Float64Array(12),
                markerWidth: markerWidth || this.defaultMarkerWidth
            };
        }
        if (markerWidth) {
            obj.markerWidth = markerWidth;
        }
        return obj;
    };

    detectMarker(image) {
        if (this._copyImageToHeap(image)) {
            return this.artoolkitNFT.artoolkitNFT.detectMarker(this.id);
        }
        return -99;
    };

    getNFTMarker(markerIndex) {
        if (0 === this.artoolkitNFT.artoolkitNFT.getNFTMarker(this.id, markerIndex)) {
            return this.artoolkitNFT.NFTMarkerInfo;
        }
    }

    detectNFTMarker() {
        this.artoolkitNFT.artoolkitNFT.detectNFTMarker(this.id);
    };

    transMatToGLMat(transMat, glMat, scale) {
        if (glMat == undefined) {
            glMat = new Float64Array(16);
        }
        glMat[0 + 0 * 4] = transMat[0]; // R1C1
        glMat[0 + 1 * 4] = transMat[1]; // R1C2
        glMat[0 + 2 * 4] = transMat[2];
        glMat[0 + 3 * 4] = transMat[3];
        glMat[1 + 0 * 4] = transMat[4]; // R2
        glMat[1 + 1 * 4] = transMat[5];
        glMat[1 + 2 * 4] = transMat[6];
        glMat[1 + 3 * 4] = transMat[7];
        glMat[2 + 0 * 4] = transMat[8]; // R3
        glMat[2 + 1 * 4] = transMat[9];
        glMat[2 + 2 * 4] = transMat[10];
        glMat[2 + 3 * 4] = transMat[11];
        glMat[3 + 0 * 4] = 0.0;
        glMat[3 + 1 * 4] = 0.0;
        glMat[3 + 2 * 4] = 0.0;
        glMat[3 + 3 * 4] = 1.0;
        if (scale != undefined && scale !== 0.0) {
            glMat[12] *= scale;
            glMat[13] *= scale;
            glMat[14] *= scale;
        }
        return glMat;
    };

    arglCameraViewRHf(glMatrix, glRhMatrix, scale) {
        var m_modelview;
        if (glRhMatrix == undefined)
            m_modelview = new Float64Array(16);
        else
            m_modelview = glRhMatrix;

        // x
        m_modelview[0] = glMatrix[0];
        m_modelview[4] = glMatrix[4];
        m_modelview[8] = glMatrix[8];
        m_modelview[12] = glMatrix[12];
        // y
        m_modelview[1] = -glMatrix[1];
        m_modelview[5] = -glMatrix[5];
        m_modelview[9] = -glMatrix[9];
        m_modelview[13] = -glMatrix[13];
        // z
        m_modelview[2] = -glMatrix[2];
        m_modelview[6] = -glMatrix[6];
        m_modelview[10] = -glMatrix[10];
        m_modelview[14] = -glMatrix[14];

        // 0 0 0 1
        m_modelview[3] = 0;
        m_modelview[7] = 0;
        m_modelview[11] = 0;
        m_modelview[15] = 1;

        if (scale != undefined && scale !== 0.0) {
            m_modelview[12] *= scale;
            m_modelview[13] *= scale;
            m_modelview[14] *= scale;
        }

        glRhMatrix = m_modelview;

        return glRhMatrix;
    };

    addEventListener(name, callback) {
        if (!this.listeners[name]) {
            this.listeners[name] = [];
        }
        this.listeners[name].push(callback);
    };

    removeEventListener(name, callback) {
        if (this.listeners[name]) {
            var index = this.listeners[name].indexOf(callback);
            if (index > -1) {
                this.listeners[name].splice(index, 1);
            }
        }
    };

    dispatchEvent(event) {
        var listeners = this.listeners[event.name];
        if (listeners) {
            for (var i = 0; i < listeners.length; i++) {
                listeners[i].call(this, event);
            }
        }
    };

   loadNFTMarkers(markerURLs, onSuccess, onError) {
        var self = this;
        this.addNFTMarkers(this.id, markerURLs, function(ids) {
            self.nftMarkerCount += ids.length;
            onSuccess(ids);
        }, onError);
    };

    loadNFTMarker(markerURL, onSuccess, onError) {
        if (markerURL) {
            this.loadNFTMarkers([markerURL], function(ids) {
              onSuccess(ids[0]);
            }, onError);
        } else {
          if (onError) {
              onError("Marker URL needs to be defined and not equal empty string!");
          }
          else {
              console.error("Marker URL needs to be defined and not equal empty string!");
          }
        }

    };
    
    addNFTMarker(arId, url, callback, onError) {
        var mId = this.nftMarkerCount++;
        var prefix = '/markerNFT_' + mId;
        var filename1 = prefix + '.fset';
        var filename2 = prefix + '.iset';
        var filename3 = prefix + '.fset3';
        this.ajax(url + '.fset', filename1, function () {
            this.ajax(url + '.iset', filename2, function () {
                this.ajax(url + '.fset3', filename3, function () {
                    var nftMarker = this.artoolkitNFT.artoolkitNFT._addNFTMarker(arId, prefix);
                    if (callback) callback(nftMarker);
                }, function (errorNumber) { if (onError) onError(errorNumber); });
            }, function (errorNumber) { if (onError) onError(errorNumber); });
        }, function (errorNumber) { if (onError) onError(errorNumber); });
    }

    addNFTMarkers(arId, urls, callback, onError) {
        var prefixes = [];
        var pending = urls.length * 3;
        var onSuccess = (filename) => {
            pending -= 1;
            if (pending === 0) {
                const vec = new this.artoolkitNFT.artoolkitNFT.StringList();
                const markerIds = [];
                for (let i = 0; i < prefixes.length; i++) {
                    vec.push_back(prefixes[i]);
                }
                var ret = this.artoolkitNFT.artoolkitNFT._addNFTMarkers(arId, vec);
                for (let i = 0; i < ret.size(); i++) {
                    markerIds.push(ret.get(i));
                }

                console.log("add nft marker ids: ", markerIds);
                if (callback) callback(markerIds);
            }
        }
        var onError = (filename, errorNumber) => {
            console.log("failed to load: ", filename);
            onError(errorNumber);
        }

        for (var i = 0; i < urls.length; i++) {
            var url = urls[i];

            var prefix = '/temp/' + url;
            prefixes.push(prefix);

            var filename1 = url + '.fset';
            var filename2 = url + '.iset';
            var filename3 = url + '.fset3';

            this.ajax(url + '.fset', filename1, onSuccess.bind(filename1), onError.bind(filename1));
            this.ajax(url + '.iset', filename2, onSuccess.bind(filename2), onError.bind(filename2));
            this.ajax(url + '.fset3', filename3, onSuccess.bind(filename3), onError.bind(filename3));
            this.nftMarkerCount += 1;
        }
    }

    ajax(url, target, callback, errorCallback) {
        callback("/temp/" + target);
    }


    setProjectionNearPlane(value) {
        return this.artoolkitNFT.artoolkitNFT.setProjectionNearPlane(this.id, value);
    };

    getProjectionNearPlane() {
        return this.artoolkitNFT.artoolkitNFT.getProjectionNearPlane(this.id);
    };

    setProjectionFarPlane(value) {
        return this.artoolkitNFT.artoolkitNFT.setProjectionFarPlane(this.id, value);
    };

    getProjectionFarPlane() {
        return this.artoolkitNFT.artoolkitNFT.getProjectionFarPlane(this.id);
    };

    async _initialize() {
        this.artoolkitNFT = await new ARToolkitNFT().init();
        console.log("[ARControllerNFT]", "ARToolkitNFT initialized");

        // load the camera
        this.cameraId = await this.artoolkitNFT.loadCamera(this.cameraParam);
        console.log("[ARControllerNFT]", "Camera params loaded with ID", this.cameraId);

        // setup
        this.id = this.artoolkitNFT.artoolkitNFT.setup(this.width, this.height, this.cameraId);
        console.log("[ARControllerNFT]", "Got ID from setup", this.id);

        this._initNFT();

        const params = this.artoolkitNFT.frameMalloc

        this.framepointer = params.framepointer;
        this.framesize = params.framesize;
        this.videoLumaPointer = params.videoLumaPointer;

        this.dataHeap = new Uint8Array(
            this.artoolkitNFT.artoolkitNFT.HEAPU8.buffer,
            this.framepointer,
            this.framesize
        );
        this.videoLuma = new Uint8Array(
            this.artoolkitNFT.artoolkitNFT.HEAPU8.buffer,
            this.videoLumaPointer,
            this.framesize / 4
        );

        this.camera_mat = new Float64Array(
            this.artoolkitNFT.artoolkitNFT.HEAPU8.buffer,
            params.camera,
            16
        );
        this.marker_transform_mat = new Float64Array(
            this.artoolkitNFT.artoolkitNFT.HEAPU8.buffer,
            params.transform,
            12
        );

        this.setProjectionNearPlane(0.1);
        this.setProjectionFarPlane(1000);

        setTimeout(() => {
            this.dispatchEvent({
                name: "load",
                target: this,
            });
        }, 1);

        return this;
    }

    _initNFT() {
        this.artoolkitNFT.artoolkitNFT.setupAR2(this.id);
    }

    _copyImageToHeap(image) {
        if (!image) {
            console.error("Error: no provided imageData to ARControllerNFT");
            return;
        }
        if (image.data) {

            var imageData = image;

        }
        //var data = imageData.data;  // this is of type Uint8ClampedArray: The Uint8ClampedArray typed array represents an array of 8-bit unsigned integers clamped to 0-255 (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8ClampedArray)
        var data = image
        //Here we have access to the unmodified video image. We now need to add the videoLuma chanel to be able to serve the underlying ARTK API
        if (this.videoLuma) {
            var q = 0;
            //Create luma from video data assuming Pixelformat AR_PIXEL_FORMAT_RGBA (ARToolKitJS.cpp L: 43)

            for (var p = 0; p < this.videoSize; p++) {
                var r = data[q + 0], g = data[q + 1], b = data[q + 2];
                // videoLuma[p] = (r+r+b+g+g+g)/6;         // https://stackoverflow.com/a/596241/5843642
                this.videoLuma[p] = (r + r + r + b + g + g + g + g) >> 3;
                q += 4;
            }
        }

        if (this.dataHeap) {
            this.dataHeap.set(data);
            return true;
        }
        return false;
    };

}
module.exports = ARControllerNFT

File: python-bindings/src/ARToolKitNFT/arcontrollerNFT.py
import ARToolKitNFT_core
import time
import asyncio

class EventDispatcher:
    def __init__(self):
        self.listeners = {}

    def add_event_listener(self, event_name, callback):
        if event_name not in self.listeners:
            self.listeners[event_name] = []
        self.listeners[event_name].append(callback)

    def remove_event_listener(self, event_name, callback):
        if event_name in self.listeners:
            self.listeners[event_name].remove(callback)

    def dispatch_event(self, event_name, event_data=None):
        if event_name in self.listeners:
            for callback in self.listeners[event_name]:
                callback(event_data)

class ARControllerNFT(EventDispatcher):
    NFT_MARKER = ARToolKitNFT_core.NFT_MARKER

    def __init__(self, width=None, height=None, cameraParam=None):
        super().__init__()
        self.id = -1
        self._width = width
        self._height = height
        self._cameraParam = cameraParam
        self.cameraId = -1

        self.artoolkitNFT = None

        self.nftMarkers = []

        self.transform_mat = [0.0] * 16
        self.transformGL_RH = [0.0] * 16
        self.camera_mat = [0.0] * 16

        self.videoWidth = width
        self.videoHeight = height
        self.videoSize = self.videoWidth * self.videoHeight if width and height else 0

        self.framesize = None
        self.videoLuma = None
        self.grayscaleEnabled = False
        self.grayscaleSource = None

        self.nftMarkerFound = False
        self.nftMarkerFoundTime = 0
        self.nftMarkerCount = 0
        self.defaultMarkerWidth = 1

        self._bwpointer = None

    def process(self, image):
        self._copyImageToHeap(image)

        for k in self.nftMarkers:
            o = self.nftMarkers[k]
            o['inPrevious'] = o['inCurrent']
            o['inCurrent'] = False

        nftMarkerCount = self.nftMarkerCount
        self.detectNFTMarker()

        MARKER_LOST_TIME = 200

        for i in range(nftMarkerCount):
            nftMarkerInfo = self.getNFTMarker(i)
            markerType = ARToolKitNFT_core.NFT_MARKER

            if nftMarkerInfo['found']:
                self.nftMarkerFound = i
                self.nftMarkerFoundTime = time.time()

                visible = self.trackNFTMarkerId(i)
                visible['matrix'] = nftMarkerInfo['pose']
                visible['inCurrent'] = True
                self.transMatToGLMat(visible['matrix'], self.transform_mat)
                self.transformGL_RH = self.arglCameraViewRHf(self.transform_mat)
                self.dispatch_event('getNFTMarker', {
                    'index': i,
                    'type': markerType,
                    'marker': nftMarkerInfo,
                    'matrix': self.transform_mat,
                    'matrixGL_RH': self.transformGL_RH,
                })
            elif self.nftMarkerFound == i:
                if time.time() - self.nftMarkerFoundTime > MARKER_LOST_TIME:
                    self.nftMarkerFound = False
                    self.dispatch_event('lostNFTMarker', {
                        'index': i,
                        'type': markerType,
                        'marker': nftMarkerInfo,
                        'matrix': self.transform_mat,
                        'matrixGL_RH': self.transformGL_RH,
                    })

    def detectNFTMarker(self):
        return self.artoolkitNFT.detectNFTMarker()

    def trackNFTMarkerId(self, id, markerWidth=None):
        obj = self.nftMarkers.get(id)
        if not obj:
            self.nftMarkers[id] = obj = {
                'inPrevious': False,
                'inCurrent': False,
                'matrix': [0.0] * 12,
                'matrixGL_RH': [0.0] * 12,
                'markerWidth': markerWidth or self.defaultMarkerWidth,
            }
        if markerWidth:
            obj['markerWidth'] = markerWidth
        return obj

    def getNFTMarker(self, markerIndex):
        return self.artoolkitNFT.getNFTMarker(markerIndex)

    def getNFTData(self, index):
        return self.artoolkitNFT.getNFTData(index)

    def debugSetup(self):
        self.setDebugMode(True)
        self._bwpointer = self.getProcessingImage()

    def transMatToGLMat(self, transMat, glMat=None, scale=None):
        if glMat is None:
            glMat = [0.0] * 16

        glMat[0] = transMat[0]
        glMat[1] = transMat[1]
        glMat[2] = transMat[2]
        glMat[3] = transMat[3]
        glMat[4] = transMat[4]
        glMat[5] = transMat[5]
        glMat[6] = transMat[6]
        glMat[7] = transMat[7]
        glMat[8] = transMat[8]
        glMat[9] = transMat[9]
        glMat[10] = transMat[10]
        glMat[11] = transMat[11]
        glMat[12] = 0.0
        glMat[13] = 0.0
        glMat[14] = 0.0
        glMat[15] = 1.0

        if scale is not None and scale != 0.0:
            glMat[12] *= scale
            glMat[13] *= scale
            glMat[14] *= scale

        return glMat

    def arglCameraViewRHf(self, glMatrix, glRhMatrix=None, scale=None):
        if glRhMatrix is None:
            glRhMatrix = [0.0] * 16

        glRhMatrix[0] = glMatrix[0]
        glRhMatrix[4] = glMatrix[4]
        glRhMatrix[8] = glMatrix[8]
        glRhMatrix[12] = glMatrix[12]
        glRhMatrix[1] = -glMatrix[1]
        glRhMatrix[5] = -glMatrix[5]
        glRhMatrix[9] = -glMatrix[9]
        glRhMatrix[13] = -glMatrix[13]
        glRhMatrix[2] = -glMatrix[2]
        glRhMatrix[6] = -glMatrix[6]
        glRhMatrix[10] = -glMatrix[10]
        glRhMatrix[14] = -glMatrix[14]
        glRhMatrix[3] = 0
        glRhMatrix[7] = 0
        glRhMatrix[11] = 0
        glRhMatrix[15] = 1

        if scale is not None and scale != 0.0:
            glRhMatrix[12] *= scale
            glRhMatrix[13] *= scale
            glRhMatrix[14] *= scale

        return glRhMatrix

    def getTransformationMatrix(self):
        return self.transform_mat

    def getCameraMatrix(self):
        return self.camera_mat

    def setDebugMode(self, mode):
        return self.artoolkitNFT.setDebugMode(mode)

    def getDebugMode(self):
        return self.artoolkitNFT.getDebugMode()

    def getProcessingImage(self):
        return self.artoolkitNFT.getProcessingImage()

    def setLogLevel(self, mode):
        return self.artoolkitNFT.setLogLevel(mode)

    def getLogLevel(self):
        return self.artoolkitNFT.getLogLevel()

    def setProjectionNearPlane(self, value):
        return self.artoolkitNFT.setProjectionNearPlane(value)

    def getProjectionNearPlane(self):
        return self.artoolkitNFT.getProjectionNearPlane()

    def setProjectionFarPlane(self, value):
        return self.artoolkitNFT.setProjectionFarPlane(value)

    def getProjectionFarPlane(self):
        return self.artoolkitNFT.getProjectionFarPlane()

    def setThresholdMode(self, mode):
        return self.artoolkitNFT.setThresholdMode(mode)

    def getThresholdMode(self):
        return self.artoolkitNFT.getThresholdMode()

    def setThreshold(self, threshold):
        return self.artoolkitNFT.setThreshold(threshold)

    def getThreshold(self):
        return self.artoolkitNFT.getThreshold()

    def setImageProcMode(self, mode):
        return self.artoolkitNFT.setImageProcMode(mode)

    def getImageProcMode(self):
        return self.artoolkitNFT.getImageProcMode()

    def setGrayData(self, data):
        self.grayscaleEnabled = True
        self.grayscaleSource = data

    def _copyImageToHeap(self, sourceImage):
        if not sourceImage:
            print("Error: no provided imageData to ARControllerNFT")
            return

        data = sourceImage.data if sourceImage.data else None

        if self.videoLuma:
            if not self.grayscaleEnabled:
                q = 0
                for p in range(self.videoSize):
                    r = data[q + 0]
                    g = data[q + 1]
                    b = data[q + 2]
                    self.videoLuma[p] = (r + r + r + b + g + g + g + g) >> 3
                    q += 4
            else:
                self.videoLuma = self.grayscaleSource

        if self.videoLuma:
            self.artoolkitNFT.passVideoData(data, self.videoLuma)
            return True

        return False

    async def _initialize(self):
        self.artoolkitNFT = ARToolKitNFT_core.ARToolKitNFT()

        print("[ARControllerNFT]", "ARToolkitNFT initialized")

        self.cameraId = await self.async_load_camera(self._cameraParam)
        print("[ARControllerNFT]", "Camera params loaded with ID", self.cameraId)

        self.id = self.artoolkitNFT.setup(self._width, self._height, self.cameraId)
        print("[ARControllerNFT]", "Got ID from setup", self.id)

        self._initNFT()

        self.framesize = self._width * self._height
        self.videoLuma = [0] * self.framesize
        self.camera_mat = self.artoolkitNFT.getCameraLens()

        self.setProjectionNearPlane(0.1)
        self.setProjectionFarPlane(1000)

        return self

    def _initNFT(self):
        self.artoolkitNFT.setupAR2()

    async def async_load_camera(self, cparam_name):
        loop = asyncio.get_event_loop()
        camera_id = await loop.run_in_executor(None, self.artoolkitNFT.loadCamera, cparam_name)
        return camera_id

    @staticmethod
    def some_static_method(arg1, arg2):
        return arg1 + arg2